clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java

```diff
@@ -1137,23 +1137,26 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
             assertEquals(partition1, partition.get());
             assertEquals(2, mockRandom.get());
 
-            // Produce large record, we should switch to next partition.
+            // Produce large record, we switched to next partition by previous produce, but
+            // for this produce the switch would be disabled because of incomplete batch.
             accum.append(topic, RecordMetadata.UNKNOWN_PARTITION, 0L, null, largeValue, Record.EMPTY_HEADERS,
                 callbacks, maxBlockTimeMs, false, time.milliseconds(), cluster);
             assertEquals(partition2, partition.get());
-            assertEquals(3, mockRandom.get());
+            assertEquals(2, mockRandom.get());
```

Comment:
The partition switching logic is now changed to switch after the batch is full, so just producing the required amount is not enough.

Reply:
Hmm, in this case, after the append, the last batch is full, which enables switching. Why isn't the partition switched?

Reply:
What happens here is the following:

1. First record (small) -- gets a new partition (because there was none)
2. Second record (large) doesn't fit, so the first record forms a batch (but not enough to switch).
3. Second record (large) creates a new batch, but it's not marked as full (disabling the switch).
4. Third record arrives, doesn't fit into the batch, it's marked as full (completing the switch, that was disabled in step 3).

So effectively in the step 4 the switch happens before the records is added, rather than after.

Reply:
Thanks, Artem. Sorry, I still don't fully understand.

After step 3, it seems that we switched to partition 2 after the append since mockRandom is 2, right? That part makes sense to me.

In step 4, we append to a new batch in partition 2. After the append(), it seems that enableSwitch should be true since `last.isFull()` should be true. Then, in `topicInfo.builtInPartitioner.updatePartitionInfo`, `producedBytes >= stickyBatchSize && enableSwitch` should be true, which will trigger partition switching. I am wondering what's missing here?

Reply:
Switching the partition only when the batch is complete make it effectively switch before we create a new batch.  In the test the step 2 seems to be an outlier, because the large record + small record seems to exceed the batch.size * 2, so the even though the switch is disabled at the end of step 2, it still happens.  I've increase the batch.size in the test, now all tests switch partition before creating a batch.